[Spring 경로 조회 - 2단계] 알린(장원영) 미션 제출합니다.

OzRagwort opened the pull request:
안녕하세요 코니 알린이에요!

이전에 남겨주신 피드백 감사합니다! 

남겨주신 피드백을 [수정](https://github.com/woowacourse/atdd-subway-path/pull/211)하고 2단계 요금 정책 추가 기능을 구현했습니다. 

잘 부탁드립니다 😄

wyRagwort committed:
docs: 추가 기능 목록 및 수정 사항 정리

wyRagwort committed:
refactor: 예상치 못한 예외가 발생했을 때 응답에 메시지가 포함되지 않도록 수정

wyRagwort committed:
refactor(Path): 최단 경로를 한번만 계산하도록 수정

wyRagwort committed:
refactor: 구간이 이어져 있지 않은 경우 404 응답을 주도록 수정

wyRagwort committed:
feat: 최단 경로 중 이용한 노선을 구하는 기능 구현

wyRagwort committed:
feat: 추가 요금이 있는 노선을 이용 할 경우 측정된 요금에 추가하는 기능 구현

wyRagwort committed:
feat: 연령별 요금 할인 기능 구현

wyRagwort committed:
fix: 하행에서 상행으로 경로를 찾지 못하던 문제 수정

wyRagwort committed:
refactor: String repeat 적용

wyRagwort committed:
refactor: 오타 수정

wyRagwort committed:
refactor: 실제 데이터와 유사한 입력으로 테스트를 수정

wyRagwort committed:
refactor: 다양한 상황에서 최단 경로를 구하는 테스트 추가

wyRagwort committed:
docs: 도메인에서 외부 라이브러리를 사용하면 발생할 수 있는 문제점 정리

wyRagwort committed:
test: 나이 구분 및 할인 정책에 대한 테스트 추가

wyRagwort committed:
refactor: 컨벤션 적용 및 불필요한 import 제거

wyRagwort committed:
docs: 노선별 추가 요금 적용 구현 완료

choihz commented on src/main/java/wooteco/subway/controller/SubwayControllerAdvice.java:
```diff
@@ -20,7 +20,7 @@ public ResponseEntity<String> handleNoSuchElementException(NoSuchElementExceptio
-        return ResponseEntity.internalServerError().body("[ERROR] " + exception.getMessage());
+    public ResponseEntity<Void> handleException() {
+        return ResponseEntity.internalServerError().build();
     }
```
> 예상하지 못한 예외를 advice에서 핸들링하는 이유가 클라이언트에게 예외에 대한 구체적인 정보를 숨기기 위해서기도 했기 때문에 메시지를 노출하면 안될 것 같습니다.
> 
> 어떤 예외가 발생할지 모르기 때문에 그냥 메시지없이 상태코드만 500으로 응답하도록 수정했습니다.

맞아요. 코멘트를 남기면서 드렸던 스크린샷의 내용처럼, 예외에서 메시지를 직접 꺼낼 경우 내부 구조 등이 노출되어 위험할 수 있지요. 따라서 예상하지 못한 예외에서는 메시지를 직접 꺼내지 않는 것이 좋습니다.

그런데, 어떤 예외가 발생할지 모르는 것은 맞으나 이 경우 대체로 서버의 오류임은 명확할 것 같아요. 그래서 "서버 오류가 발생했습니다." 등 간단한 메시지 정도는 줘도 괜찮을 것 같아요.

OzRagwort replied:
간단한 메시지 정도는 괜찮을 것 같네요! 수정했습니다.

choihz commented on src/main/java/wooteco/subway/domain/Path.java:
```diff
@@ -1,38 +1,22 @@
-    public int getShortestDistance(Station source, Station target) {
-        DijkstraShortestPath<Station, DefaultWeightedEdge> dijkstraShortestPath = getDijkstraShortestPath();
+    private final List<Station> stations;
+    private final int distance;
```
`Path`라는 이름이 지나치게 추상적이지 않나 하는 생각이 드는데요, 이 객체는 어떤 정보를 담고 있나요?

OzRagwort replied:
처음 Path는 최단 경로, 최단 거리를 구할 수 있도록 생각했습니다.

`DijkstraShortestPathStation`를 분리하면서 도메인보다는 DTO처럼 느껴질 정도로 하는 일이 없어졌다가 수정하면서 최단 경로, 최단 거리, 경로와 나이에 따른 요금 계산을 처리하도록 수정되었습니다.

최단 경로, 최단 거리, 경로와 나이에 따른 요금 계산을 모두 하기 때문에 공통점인 경로라는 키워드를 사용하여 Path라는 이름을 가지게 되었습니다.

choihz commented on src/main/java/wooteco/subway/utils/DijkstraShortestPathStation.java:
```diff
@@ -0,0 +1,59 @@
+package wooteco.subway.utils;
```
**유틸**이란 무엇일까요?

OzRagwort replied:
유틸 클래스는 여러 곳에서 범용적으로 사용될 수 있는 클래스를 의미한다고 생각합니다.

하지만 `DijkstraShortestPathStation`는 PathService에서만 사용되기 때문에 유틸이라고 보기 어려울 것 같습니다.

이 부분을 도메인에 넣지 않기위해 전략 패턴으로 만들어봤습니다.

외부 라이브러리는 버전에 따라 필요에 따라 수정이 될 수 있다고 생각합니다. 버전이 올라가면서 호환이 되지 않을 수도 있고 로직이 변경되면서 다른 라이브러리를 사용할 수도 있을 것입니다.

이런 상황에서 좀 더 유연하게 외부 라이브러리를 적용할 수 있도록 인터페이스를 만들고 DijkstraShortestPath를 사용하는 전략을 만들었습니다.

choihz commented on src/main/java/wooteco/subway/utils/DijkstraShortestPathStation.java:
```diff
@@ -0,0 +1,59 @@
+                .filter(station -> station.getId().equals(stationId))
+                .findFirst()
+                .orElseThrow(() -> new NoSuchElementException("역이 없습니다."));
+    }
```
사용하지 않는 것으로 보이는군요 😢

OzRagwort replied:
제거했습니다!

choihz commented on src/main/java/wooteco/subway/utils/DijkstraShortestPathStation.java:
```diff
@@ -0,0 +1,59 @@
+import wooteco.subway.domain.Sections;
+import wooteco.subway.domain.Station;
+
+public class DijkstraShortestPathStation {
```
`DijkstraShortestPathStation`은 객체로서 존재하지 못하고 외부에 함수만 빌려주는 친구인데요, 이런 방식으로 인해 객체지향과는 다소 거리가 생긴 것 같아요. 중요한 일을 얘가 다 하고 있어서 `Path`는 하는 일이 하나도 없고요. 알린이 이런 방식을 채택한 이유는 무엇일까요? 어떤 장점을 느끼셨나요?

OzRagwort replied:
다시 천천히 코드를 보니 DijkstraShortestPathStation도 유틸 클래스로 보기 어렵고 Path도 하는 일이 없고 도메인이라고 보기 어려울 것 같았습니다. 그래서 느껴지는 장점은 없는 것 같습니다.

뭔가 외부 라이브러리를 분리해보자는 것에 집중하다보니 극단적으로 수정이 된 것 같습니다.

외부 라이브러리가 모든 일을 하지 않고 도메인이 할 일은 Path가 가지도록 수정했습니다. (거리에 따른 요금 계산과 연령별 할인 처리)

choihz commented on src/main/java/wooteco/subway/service/PathService.java:
```diff
@@ -1,30 +1,42 @@
+        this.lineDao = lineDao;
     }
 
     public PathServiceResponse getShortestPath(PathServiceRequest pathServiceRequest) {
```
20여 라인의 대형 메서드가 되고 말았네요 😭 역할을 좀 더 나누고 이해하기 쉽게 개선해 볼 수는 없을까요?

OzRagwort replied:
도메인으로 이동할 부분은 이동하고 메서드 분리를 할 부분은 분리하여 개선해보았습니다 !

choihz commented on src/main/java/wooteco/subway/service/PathService.java:
```diff
@@ -34,14 +46,64 @@ public PathServiceResponse getShortestPath(PathServiceRequest pathServiceRequest
+            maxExtraFare = Math.max(maxExtraFare, lineEntity.getExtraFare());
+        }
+        return maxExtraFare;
     }
```
이런 로직들이 `Service`에 위치하는 것이 적절한가요?

OzRagwort replied:
이 두 메서드는 Sections, Section, Path, Lines, Line 등 도메인으로 이동시켰습니다 !

choihz commented on src/test/java/wooteco/subway/utils/DijkstraShortestPathStationTest.java:
```diff
@@ -0,0 +1,142 @@
+        // given
+        Station station1 = new Station(1L, "name1");
+        Station station2 = new Station(2L, "name2");
+        Station station3 = new Station(3L, "name3");
```
여기는 놓치셨네용 🤓

OzRagwort replied:
수정했습니다! 꼼꼼하게 확인하지 못했네요 ㅠㅠ

choihz commented on src/test/java/wooteco/subway/utils/DijkstraShortestPathStationTest.java:
```diff
@@ -0,0 +1,142 @@
+                .containsExactly(station1, station3);
+    }
+
+    @DisplayName("최단 거리가 같다면 역이 더 적은 경로로 최단 경로를 구한다.")
```
👍

choihz commented on src/test/java/wooteco/subway/utils/DijkstraShortestPathStationTest.java:
```diff
@@ -0,0 +1,142 @@
+                .containsExactly(station1, station2, station5);
+    }
+
+    @DisplayName("두 역이 이어져있지 않으면 ")
```
⁉️

OzRagwort replied:
ㅠㅠ... 더 신경써서 작성해보겠습니다.

choihz commented on src/main/java/wooteco/subway/utils/DijkstraShortestPathStation.java:
```diff
@@ -0,0 +1,59 @@
+
+    public static Path getPath(Sections sections, Station source, Station target) {
+        DijkstraShortestPath<Station, DefaultWeightedEdge> dijkstraShortestPath = getDijkstraShortestPath(sections);
+        GraphPath<Station, DefaultWeightedEdge> graphPath = dijkstraShortestPath.getPath(source, target);
```
<img width="1076" alt="스크린샷 2022-05-23 오전 2 10 17" src="https://user-images.githubusercontent.com/18049492/169707220-a5d46cdf-a810-40f4-aabc-9f00552ceb7e.png">

1 - 2만 연결된 상태에서 1 - 3의 경로를 조회하려 하니 여기서 터지는 것 같네요 👀

OzRagwort replied:
1-2 3-4만 연결된 상태에서 1-3으로 경로를 조회하는 경우는 체크를 했는데 아예 역이 구간에 등록되지 않은 상태에서는 확인하지 못했습니다. 

이 경우에 대한 테스트를 만들고 수정했습니다!

choihz commented on src/main/java/wooteco/subway/service/PathService.java:
```diff
@@ -34,14 +46,64 @@ public PathServiceResponse getShortestPath(PathServiceRequest pathServiceRequest
                 .collect(Collectors.toList());
-        int distance = path.getShortestDistance(source, target);
-        FareStrategy fare = new DistanceFareStrategy();
+        List<Long> visitLines = getVisitLines(path.getStations(), sectionEntities);
```
출발역과 도착역을 같게 하면 어떤 일이 일어나고 있나요?

OzRagwort replied:
경로가 1개의 역이 나오게 되는데 이 경우 거리도 0이고 요금 계산도 할 수 없습니다. 

제자리로 이동하려는 경우를 검증하여 예외를 발생시키도록 수정했습니다.

choihz commented on src/main/java/wooteco/subway/service/DomainCreatorService.java:
```diff
@@ -27,13 +26,6 @@ public class DomainCreatorService {
-        return new Path(new Sections(sections));
-    }
-
     Line createLine(Long lineId) {
```
이번 미션의 내용은 아니기 때문에 꼭 고치실 필요는 없지만, 참고하시라고 남깁니다.

<img width="401" alt="스크린샷 2022-05-23 오전 2 18 06" src="https://user-images.githubusercontent.com/18049492/169707543-67f67eed-e2ea-4a7a-852d-367f5d2370d1.png">  <img width="169" alt="스크린샷 2022-05-23 오전 2 18 23" src="https://user-images.githubusercontent.com/18049492/169707552-f848e3d0-958a-4604-815f-556a53f31957.png">
<img width="526" alt="스크린샷 2022-05-23 오전 2 18 30" src="https://user-images.githubusercontent.com/18049492/169707562-392ccfa3-db40-4ebe-99e5-49b0dc8b0f07.png">

제공된 프론트엔드 페이지의 구간 관리 메뉴에 접속해 보면, 구간이 순서대로 노출되지 않고 있네요.

OzRagwort replied:
가장 상행역부터 하행 역까지 순서대로 출력되도록 수정했습니다.

choihz commented on src/main/java/wooteco/subway/service/PathService.java:
```diff
@@ -1,30 +1,42 @@
+        this.lineDao = lineDao;
     }
 
     public PathServiceResponse getShortestPath(PathServiceRequest pathServiceRequest) {
```
`source`, `target`, `age` 중 들어오지 않은 값이 있으면 어떻게 되나요?

OzRagwort replied:
위 값은 모두 `@RequestParam`으로 받습니다. `@RequestParam`은 기본적으로 required 옵션이 true이고 무조건 값이 있어야 합니다.

따라서 source, target, age 중 어떤 하나라도 빠지면 `MissingServletRequestParameterException` 예외가 발생합니다. 

`MissingServletRequestParameterException`를 Advice에서 처리하도록 수정했습니다.

choihz reviewed (changes requested):
안녕하세요, 알린!

지하철 경로 조회 미션의 2단계도 잘 구현해 주셨네요 👍 몇 가지 피드백을 남겨 두었는데요, 보시고 궁금한 점 또는 의견이 있으시다면 언제든 코멘트 또는 DM 주세요~

wyRagwort committed:
docs: 2 단계 피드백, 수정사항 정리

wyRagwort committed:
refactor: 잘못된 메시지, 변수 수정

wyRagwort committed:
refactor: 출발역과 도착역이 같은 경우 에러가 발생하도록 수정

wyRagwort committed:
refactor: 역이 어느 구간에도 포함되어 있지 않은 경우 에러가 발생하도록 수정

wyRagwort committed:
refactor: @RequestParam의 필수 요소를 입력하지 않을 때 예외를 핸들링하도록 수정

wyRagwort committed:
fix: 테스트 given 오류 수정

wyRagwort committed:
refactor: 노선의 모든 역을 상행역부터 하행역까지 순서대로 정렬하여 응답하도록 수정

wyRagwort committed:
refactor(Path): 메서드, 역할 분리, 코드 개선

- 경로 탐색 라이브러리를 전략 패턴으로 만들어 사용하도록 수정
- 도메인이 가져야 할 코드를 도메인으로 이동
- 메서드 분리 등으로 코드 가독성 개선
- 사용하지 않는 메서드 제거

wyRagwort committed:
test: 예외가 발생할 수 있는 상황에 대한 테스트 추가

wyRagwort committed:
refactor: 예상하지 못한 예외에 대한 응답 시 void 대신 간단한 안내 문구를 응답하도록 수정

wyRagwort committed:
chore: spring-boot-starter-validation 추가

wyRagwort committed:
refactor: Controller 요청 객체에서 빈 값이 들어오지 않도록 검증 추가

choihz commented on src/main/java/wooteco/subway/service/PathService.java:
```diff
@@ -1,52 +1,104 @@
-    private final DomainCreatorService domainCreatorService;
+    private final SectionDao sectionDao;
+    private final LineDao lineDao;
+    private final DijkstraPathStrategy pathStrategy;
```
이렇게 되면 꼭 다익스트라 알고리즘만을 사용해야 하겠네요 ㅎㅎ

choihz commented on src/main/java/wooteco/subway/strategy/DijkstraShortestPathStrategy.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    private static void validateExistInSections(Sections sections, Station station) {
+        if (!sections.getStations().contains(station)) {
+            throw new NoSuchElementException("이동할 수 있는 경로가 없습니다.");
+        }
```
`Sections`에서 역을 직접 가져와 확인하기보다, `Sections`에게 요청하도록 만들 수도 있겠군요 👀

choihz commented on src/main/java/wooteco/subway/service/PathService.java:
```diff
@@ -1,52 +1,104 @@
     public PathServiceResponse getShortestPath(PathServiceRequest pathServiceRequest) {
         validateExists(pathServiceRequest.getSource());
         validateExists(pathServiceRequest.getTarget());
+        validateNotSameStation(pathServiceRequest.getSource(), pathServiceRequest.getTarget());
```
동일 역 여부 검증은 `PathService`보다는 `PathStrategy.getPath()` 내부에서 직접 하는 것이 더 나을 것 같은데요, 왜 그럴까요?

choihz commented on src/main/java/wooteco/subway/service/PathService.java:
```diff
@@ -1,52 +1,104 @@
+                path.getFare(pathServiceRequest.getAge()));
+    }
+
+    private Lines getAllLines() {
+        List<LineEntity> lineEntities = lineDao.findAll();
+        List<SectionEntity> sectionEntities = sectionDao.findAll();
+
+        List<Line> lines = new ArrayList<>();
+        for (LineEntity lineEntity : lineEntities) {
+            lines.add(new Line(
+                    lineEntity.getId(),
+                    lineEntity.getName(),
+                    lineEntity.getColor(),
+                    lineEntity.getExtraFare(),
+                    getSectionsByLine(sectionEntities, lineEntity.getId())
+            ));
+        }
+
+        return new Lines(lines);
+    }
```
이런 식의 변환 로직이 지금 너무 많은데, 여기에 위치하는 것이 적절한지, 더 나은 방법은 없을지 고민해 보면 좋을 것 같아요.

choihz commented on src/main/java/wooteco/subway/service/PathService.java:
```diff
@@ -1,52 +1,104 @@
+        List<StationDto> stationDtos = stations.stream()
                 .map(station -> new StationDto(station.getId(), station.getName()))
                 .collect(Collectors.toList());
-        int distance = path.getShortestDistance(source, target);
-        FareStrategy fare = new DistanceFareStrategy();
-
-        return new PathServiceResponse(stationDtos, distance, fare.calculate(distance));
+        return stationDtos;
     }
```
변수 할당 없이 바로 리턴해도 되겠죠?

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -98,6 +105,45 @@ public List<Station> getStations() {
+                .orElseThrow(() -> new IllegalStateException("구간 정보가 잘못되었습니다."));
+    }
+
+    private List<Station> getUpStations() {
+        return sections.stream()
+                .map(Section::getUpStation)
+                .collect(Collectors.toList());
+    }
+
+    private List<Station> getDownStations() {
+        return sections.stream()
+                .map(Section::getDownStation)
+                .collect(Collectors.toList());
+    }
```
이 메서드들은 `getStations()`에서도 사용할 수 있겠네요~

choihz commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -35,6 +38,85 @@ void createLine() {
+        );
+    }
+
+    @DisplayName("이름의 길이가 256 이상이면 노선을 만들 수 없다.")
```
꼼꼼하게 테스트하셨군요 👍

다만 조금 더 고민해 보면 좋을 부분이 있어요. 이런 케이스를 모두 **인수 테스트**로 검증해야만 할까요? 인수 테스트의 목적은 무엇인가요? 그리고, 인수 테스트의 비용은 어떻게 되나요?

choihz commented on src/main/java/wooteco/subway/domain/Path.java:
```diff
@@ -1,38 +1,31 @@
+    }
 
-        return new DijkstraShortestPath<>(graph);
+    public int getFare(int age) {
+        int fare = new DistanceFareStrategy().calculate(distance) + visitLines.getMaxExtraFare();
+        AgePolicy agePolicy = AgePolicy.fromAge(age);
+        return agePolicy.getDiscountedFare(fare);
     }
```
`Path`라는 객체가 `age`를 받아 요금을 계산하고 있는데, 이게 과연 `Path`의 책임이어야만 할까요?

choihz reviewed (approved):
안녕하세요, 알린!

피드백 잘 반영해 주신 것 확인했습니다 👍 이제 이번 미션을 마무리해도 될 것 같아요. 다만, 더 생각해 보면 좋을 부분들에 코멘트를 달아 두었으니 확인해 주시고, 혹시나 궁금한 점이 있으시다면 코멘트 또는 DM 주세요!

그동안 정말 고생 많으셨고, 다음에 다시 만나요~ 👋
